### PR TITLE
[Distributed] Protocol mangling and primary associated type fixes

### PR DESCRIPTION
**Pull request by ktoso**
The commits are now properly staged, please check their descriptions for details.

This solves:
- rdar://139332556 - so now we're handling primary associated types in protocols with @resolvable -- and this would have then crashed during mangling (!), which is why the follow up commits
- rdar://139781083 - which is solved in [`2a3ac44` (#77584)](https://github.com/swiftlang/swift/pull/77584/commits/2a3ac4458fcca51906ffd53a40129347ed8d81d7) which was about how sometimes with generics involved or more complex type signatures of a `distributed func` we'd wrongly mangle and crash during thunk mangling.

Both changes go hand in hand and allow correct handling of more situations with generics and complex types in `distributed func`s

### CODE REVIEW

**Comment by xedin** [comment]
@ktoso I see that there is a bunch of leftover comments and code, could you please clean that up and ping me when done so I can take a look when this is ready to merge?

**Comment by ktoso** [comment]
Oh yeah don't review this at all yet, I'm cleaning up and will make it into some nice commits.

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Unrelated, Cxx interop fails compiling on Windows:

```
inalloca argument for call has mismatched alloca

  %7 = alloca %TSo3stdO0077basic_stringCChar32char_traitsCChar32allocatorCChar32_jlHEvsaGGdtaEBguasEitaaV, align 4

  %9 = call i32 @"?__swift_interopComputeHashOfU32String@@YAIV?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@@Z"(ptr %7) #16

<unknown>:0: error: fatal error encountered during compilation; please submit a bug report (https://swift.org/contributing/#reporting-bugs)

```

**Comment by ktoso** [comment]
@swift-ci please smoke test Windows

File: test/Distributed/Runtime/distributed_actor_remoteCall_protocol_method_in_presence_of_multiple_systems.swift [hunk 1]
```diff
@@ -46,16 +67,18 @@ distributed actor DAFL: GreeterProtocol {
 
     print("resolved on \(fakeRoundtripSystem): \(type(of: gfr))")
     // CHECK: resolved on main.FakeRoundtripActorSystem: $GreeterProtocol<FakeRoundtripActorSystem>
-
-    // CHECK: > execute distributed target: main.$GreeterProtocol.greet(), identifier: $s4main16$GreeterProtocolC5greetSSyYaKFTE
+    // CHECK-NEXT: > encode generic sub: main.$GreeterProtocol<main.FakeRoundtripActorSystem>
+    // CHECK-NEXT: > encode return type: Swift.String
+    // CHECK-NEXT: > done recording
+    // CHECK-NEXT: >> remoteCall: on:main.$GreeterProtocol<main.FakeRoundtripActorSystem>, target:main.$GreeterProtocol.greet(), invocation:FakeInvocationEncoder(genericSubs: [main.$GreeterProtocol<main.FakeRoundtripActorSystem>], arguments: [], returnType: Optional(Swift.String), errorType: nil), throwing:Swift.Never, returning:Swift.String
+    // CHECK-NEXT: > execute distributed target: main.$GreeterProtocol.greet(), identifier: $s4main16$GreeterProtocolC5greetSSyYaKFTE
```
Comment (ktoso) [inline test/Distributed/Runtime/distributed_actor_remoteCall_protocol_method_in_presence_of_multiple_systems.swift:74]: note: here we keep the same mangling, good 👍

File: test/Distributed/Runtime/distributed_actor_func_calls_remoteCall_extension.swift [hunk 2]
```diff
@@ -40,7 +40,9 @@ func test() async throws {
   let ref = try KappaProtocolImpl.resolve(id: local.id, using: system)
 
   let reply = try await ref.echo(name: "Caplin")
-  // CHECK: >> remoteCall: on:main.KappaProtocolImpl, target:main.$KappaProtocol.echo(name:), invocation:FakeInvocationEncoder(genericSubs: [main.KappaProtocolImpl], arguments: ["Caplin"], returnType: Optional(Swift.String), errorType: nil), throwing:Swift.Never, returning:Swift.String
+  // CHECK: >> remoteCall: on:main.KappaProtocolImpl, target:echo(name:), invocation:FakeInvocationEncoder(genericSubs: [main.KappaProtocolImpl], arguments: ["Caplin"], returnType: Optional(Swift.String), errorType: nil), throwing:Swift.Never, returning:Swift.String
+  // CHECK: > execute distributed target: echo(name:), identifier: $s4main13KappaProtocolPAAE4echo4nameS2S_tYaKFTE
```
Comment (ktoso) [inline test/Distributed/Runtime/distributed_actor_func_calls_remoteCall_extension.swift:44]: added a mangling check as well, we didnt before

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/AST/ASTMangler.cpp [hunk 3]
```diff
@@ -4452,12 +4452,32 @@ void ASTMangler::appendDistributedThunk(
...
+  // recipient, whichever 'protocol Type'-conforming type it will be.
+  NominalTypeDecl *stubActorDecl = nullptr;
+  if (auto P = referenceInProtocolContextOrRequirement()) {
+    auto &C = thunk->getASTContext();
+    auto M = thunk->getModuleContext();
+
+    SmallVector<ValueDecl *, 1> stubClassLookupResults;
+    C.lookupInModule(M, ("$" + P->getNameStr()).str(), stubClassLookupResults);
+
+    if (stubClassLookupResults.size() > 0) {
```
Comment (xedin) [inline lib/AST/ASTMangler.cpp:4468]: I think we need to make sure that it's exactly one here via an ASSERT.

File: lib/AST/ASTMangler.cpp [hunk 4]
```diff
@@ -4452,12 +4452,32 @@ void ASTMangler::appendDistributedThunk(
...
+  // does not know the type of the recipient distributed actor, and we use the
+  // $Target type as substitute to then generically invoke it on the type of the
+  // recipient, whichever 'protocol Type'-conforming type it will be.
+  NominalTypeDecl *stubActorDecl = nullptr;
+  if (auto P = referenceInProtocolContextOrRequirement()) {
+    auto &C = thunk->getASTContext();
+    auto M = thunk->getModuleContext();
+
+    SmallVector<ValueDecl *, 1> stubClassLookupResults;
+    C.lookupInModule(M, ("$" + P->getNameStr()).str(), stubClassLookupResults);
```
Comment (xedin) [inline lib/AST/ASTMangler.cpp:4466]: I think you can use `llvm::Twine('$', P->getNameStr()).str()` might be slightly better here

**Review by xedin — APPROVED** [review]
LGTM! Could you add an example with array type and optional as parameters to make sure that it works as expected too?

File: lib/AST/ASTMangler.cpp (hunk 3 above)
Comment (ktoso) [reply lib/AST/ASTMangler.cpp:4468]: Weirdly enough I think i've seen multiple here... I'll see what I can do

File: lib/AST/ASTMangler.cpp (hunk 4 above)
Comment (ktoso) [reply lib/AST/ASTMangler.cpp:4466]: Hm I think the + forms a Twine actually but I can be explicit about it yeap

**Comment by ktoso** [comment]
Thanks for review! Huh seems I missed to git add my array test :) Adding some more

File: lib/AST/ASTMangler.cpp (hunk 4 above)
Comment (xedin) [reply lib/AST/ASTMangler.cpp:4466]: It would form std::string

File: lib/AST/ASTMangler.cpp (hunk 3 above)
Comment (xedin) [reply lib/AST/ASTMangler.cpp:4468]: That won’t be unexpected though right? There should be only one $ thing per protocol?

Comment (ktoso) [reply lib/AST/ASTMangler.cpp:4468]: Yeah not sure how multiple can happen, will dig into it

File: lib/AST/ASTMangler.cpp (hunk 4 above)
Comment (ktoso) [reply lib/AST/ASTMangler.cpp:4466]: Ah here I guess so, I misremembered some other piece of code I think. Thanks, fixed

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/AST/ASTMangler.cpp (hunk 3 above)
Comment (ktoso) [reply lib/AST/ASTMangler.cpp:4470]: Seems I was seeing things, assert is fine -- added it back

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Pull request merged**